AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -997,6 +1028,7 @@ - (void)rangeController:(ASRangeController *)rangeController didDeleteNodes:(NSA
     [_layoutFacilitator collectionViewWillEditCellsAtIndexPaths:indexPaths batched:NO];
     [UIView performWithoutAnimation:^{
       [super deleteItemsAtIndexPaths:indexPaths];
```

Comment:
To point 1: Look into `rangeController:didEndUpdatesAnimated:completion:` there you see we call `_scheduleCheckForBatchFetching` in the completion handler. This should be the behavior you mention in your comment, it's deferring the check for batch fetching until all the `_batchUpdateBlocks` have finished.